src/group_theory/sylow.lean

```diff
@@ -17,21 +17,146 @@ The Sylow theorems are the following results for every finite group `G` and ever
+* `is_p_group.exists_le_sylow`: A generalization of Sylow's first theorem:
+  Every `p`-subgroup is contained in a Sylow `p`-subgroup.
+* `sylow_conjugate`: A generalization of Sylow's second theorem:
+  If the number of Sylow `p`-subgroups is finite, then all Sylow `p`-subgroups are conjugate.
+* `card_sylow_modeq_one`: A generalization of Sylow's third theorem:
+  If the number of Sylow `p`-subgroups is finite, then it is congruent to `1` modulo `p`.
+-/
+
+open fintype mul_action subgroup
 
-## TODO
+section infinite_sylow
 
-* Prove the second and third of the Sylow theorems.
-* Sylow theorems for infinite groups
--/
+variables (p : ℕ) (G : Type*) [group G]
+
+/-- A Sylow `p`-subgroup is a maximal `p`-subgroup. -/
+structure sylow extends subgroup G :=
+(is_p_group' : is_p_group p to_subgroup)
+(is_maximal' : ∀ {Q : subgroup G}, is_p_group p Q → to_subgroup ≤ Q → Q = to_subgroup)
+
+variables {p} {G}
+
+instance : has_coe (sylow p G) (subgroup G) := ⟨sylow.to_subgroup⟩
+
+@[ext] lemma sylow.ext {P Q : sylow p G} (h : (P : subgroup G) = Q) : P = Q :=
+by cases P; cases Q; congr'
+
+lemma sylow.ext_iff {P Q : sylow p G} : P = Q ↔ (P : subgroup G) = Q :=
+⟨congr_arg coe, sylow.ext⟩
+
+/-- A generalization of **Sylow's first theorem**.
+  Every `p`-subgroup is contained in a Sylow `p`-subgroup. -/
+lemma is_p_group.exists_le_sylow {P : subgroup G} (hP : is_p_group p P) :
+  ∃ Q : sylow p G, P ≤ Q :=
+exists.elim (zorn.zorn_nonempty_partial_order₀ {Q : subgroup G | is_p_group p Q} (λ c hc1 hc2 Q hQ,
+⟨ { carrier := ⋃ (R : c), R,
+    one_mem' := ⟨Q, ⟨⟨Q, hQ⟩, rfl⟩, Q.one_mem⟩,
+    inv_mem' := λ g ⟨_, ⟨R, rfl⟩, hg⟩, ⟨R, ⟨R, rfl⟩, R.1.inv_mem hg⟩,
+    mul_mem' := λ g h ⟨_, ⟨R, rfl⟩, hg⟩ ⟨_, ⟨S, rfl⟩, hh⟩, (hc2.total_of_refl R.2 S.2).elim
+      (λ T, ⟨S, ⟨S, rfl⟩, S.1.mul_mem (T hg) hh⟩) (λ T, ⟨R, ⟨R, rfl⟩, R.1.mul_mem hg (T hh)⟩) },
+  λ ⟨g, _, ⟨S, rfl⟩, hg⟩, by
+  { refine exists_imp_exists (λ k hk, _) (hc1 S.2 ⟨g, hg⟩),
+    rwa [subtype.ext_iff, coe_pow] at hk ⊢ },
+  λ M hM g hg, ⟨M, ⟨⟨M, hM⟩, rfl⟩, hg⟩⟩) P hP) (λ Q ⟨hQ1, hQ2, hQ3⟩, ⟨⟨Q, hQ1, hQ3⟩, hQ2⟩)
+
+instance sylow.nonempty : nonempty (sylow p G) :=
+nonempty_of_exists is_p_group.of_bot.exists_le_sylow
+
+noncomputable instance sylow.inhabited : inhabited (sylow p G) :=
+classical.inhabited_of_nonempty sylow.nonempty
+
+open_locale pointwise
+
+/-- `subgroup.pointwise_mul_action` preserves Sylow subgroups. -/
+instance sylow.pointwise_mul_action {α : Type*} [group α] [mul_distrib_mul_action α G] :
+  mul_action α (sylow p G) :=
+{ smul := λ g P, ⟨g • P, P.2.map _, λ Q hQ hS, inv_smul_eq_iff.mp (P.3 (hQ.map _)
+    (λ s hs, (congr_arg (∈ g⁻¹ • Q) (inv_smul_smul g s)).mp
+      (smul_mem_pointwise_smul (g • s) g⁻¹ Q (hS (smul_mem_pointwise_smul s g P hs)))))⟩,
+  one_smul := λ P, sylow.ext (one_smul α P),
+  mul_smul := λ g h P, sylow.ext (mul_smul g h P) }
+
+instance sylow.mul_action : mul_action G (sylow p G) :=
+mul_action.comp_hom _ mul_aut.conj
```

Comment:
This overlaps somewhat with #8592.

Reply:
Even once #8592 is merged, I think this would still be a reasonable approach. I don't think that #8592 makes this any easier?

Reply:
I suppose #8592 might make `sylow.smul_eq_iff_mem_normalizer` a bit easier.